tests/end_to_end_test/end_to_end_test.go

```diff
@@ -553,6 +554,149 @@ func TestSnapshotDeleteTypeCheck(t *testing.T) {
 	e.runAndExpectFailure(t, "snapshot", "delete", manifestID, "--unsafe-ignore-source")
 }
 
+func TestSnapshotFail(t *testing.T) {
+	e := newTestEnv(t)
+	defer e.cleanup(t)
+	defer e.runAndExpectSuccess(t, "repo", "disconnect")
+
+	e.runAndExpectSuccess(t, "repo", "create", "filesystem", "--path", e.repoDir)
+	e.runAndExpectSuccess(t, "policy", "set", "--global", "--keep-latest", strconv.Itoa(1<<31-1))
+	source := filepath.Join(e.dataDir, "source")
+
+	// Test snapshot of nonexistent directory fails
+	e.runAndExpectFailure(t, "snapshot", "create", source)
+
+	testDirDepth := 2
+	createDirectory(t, source, testDirDepth)
```

Comment:
You may want to construct a more _constrained_ source directory for the test. Otherwise, this ends up performing the same test multiple times. The size of the tree grows exponentially `O(5^N)`.
So you end up performing the same check for files many times, same for directories.
Ultimately, you want to reduce the test cases to
  - once for a directory
  - once for a file
and then perform the appropriate permutations of the permissions. Even those could be trimmed down
to the 4 cases that are relevant.
- `00` : `___` 
- `01` : `__x`
- `04` : `r__`
- `05` : `r_x`

Reply:
the limitations for once per directory and once per file are done in the `testPermissions` helper.
Even so I actually implemented the smaller test directory just before you posted the review (https://github.com/kastenhq/kopia/pull/20/commits/9beac9b47d287adbf1c13202a83299631c11583e)

Reply:
Good suggestion on cutting down the permissions iterations; I'll trim that down